x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialCentroid.java

```diff
@@ -0,0 +1,92 @@
+import org.elasticsearch.compute.aggregation.SpatialCentroidGeoPointDocValuesAggregatorFunctionSupplier;
+import org.elasticsearch.compute.aggregation.SpatialCentroidGeoPointSourceValuesAggregatorFunctionSupplier;
+import org.elasticsearch.xpack.esql.EsqlIllegalArgumentException;
+import org.elasticsearch.xpack.esql.expression.function.FunctionInfo;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.esql.planner.ToAggregator;
+import org.elasticsearch.xpack.esql.type.EsqlDataTypes;
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.expression.function.aggregate.SpatialAggregateFunction;
+import org.elasticsearch.xpack.ql.tree.NodeInfo;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+
+import java.util.List;
+
+import static org.elasticsearch.xpack.esql.expression.EsqlTypeResolutions.isSpatial;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.DEFAULT;
+
+/**
+ * Calculate spatial centroid of all geo_point or cartesian point values of a field in matching documents.
+ */
+public class SpatialCentroid extends SpatialAggregateFunction implements ToAggregator {
+
+    @FunctionInfo(returnType = { "geo_point", "cartesian_point" }, description = "The centroid of a spatial field.", isAggregation = true)
+    public SpatialCentroid(Source source, @Param(name = "field", type = { "geo_point", "cartesian_point" }) Expression field) {
+        super(source, field, false);
+    }
+
+    private SpatialCentroid(Source source, Expression field, boolean useDocValues) {
+        super(source, field, useDocValues);
+    }
+
+    @Override
+    public SpatialCentroid withDocValues() {
+        return new SpatialCentroid(source(), field(), true);
+    }
+
+    @Override
+    protected Expression.TypeResolution resolveType() {
+        return isSpatial(field(), sourceText(), DEFAULT);
+    }
+
+    @Override
+    public DataType dataType() {
+        // We aggregate incoming GEO_POINTs into a single GEO_POINT, or incoming CARTESIAN_POINTs into a single CARTESIAN_POINT.
+        return field().dataType();
+    }
+
+    @Override
+    protected NodeInfo<SpatialCentroid> info() {
+        return NodeInfo.create(this, SpatialCentroid::new, field());
+    }
+
+    @Override
+    public SpatialCentroid replaceChildren(List<Expression> newChildren) {
+        return new SpatialCentroid(source(), newChildren.get(0));
+    }
+
+    @Override
+    public AggregatorFunctionSupplier supplier(List<Integer> inputChannels) {
+        DataType type = field().dataType();
+        if (useDocValues) {
+            // When the points are read as doc-values (eg. from the index), feed them into the doc-values aggregator
+            if (type == EsqlDataTypes.GEO_POINT) {
+                return new SpatialCentroidGeoPointDocValuesAggregatorFunctionSupplier(inputChannels);
```

Comment:
Look at the changes in `AggregateMapper` to see how we supported these combinations of generated function names.